tensorflow/core/kernels/decode_bmp_op.cc

```diff
@@ -0,0 +1,125 @@
+#include "tensorflow/core/framework/types.h"
+#include "tensorflow/core/framework/types.pb.h"
+#include "tensorflow/core/lib/core/status.h"
+#include "tensorflow/core/platform/logging.h"
+
+namespace tensorflow {
+
+// Decode the contents of a BMP file
+class DecodeBmpOp : public OpKernel {
+ public:
+  explicit DecodeBmpOp(OpKernelConstruction* context) : OpKernel(context) {}
+
+  void Compute(OpKernelContext* context) override {
+    const Tensor& contents = context->input(0);
+    OP_REQUIRES(context, TensorShapeUtils::IsScalar(contents.shape()),
+                errors::InvalidArgument("contents must be scalar, got shape ",
+                                        contents.shape().DebugString()));
+
+    // Start decoding image to get shape details
+    const StringPiece input = contents.scalar<string>()();
+
+    // Decode image, allocating tensor once the image size is known
+    Tensor* output = nullptr;
+    OP_REQUIRES(
+        context,
+        Decode(input.data(), input.size(),
+                    [=, &output](int width, int height,
+                                 int channels) -> uint8* {
+                      Status status(context->allocate_output(
+                          0, TensorShape({height, width, channels}),
+                          &output));
+                      if (!status.ok()) {
+                        VLOG(1) << status;
+                        context->SetStatus(status);
+                        return nullptr;
+                      }
+                      return output->flat<uint8>().data();
+                    }),
+        errors::InvalidArgument("Invalid BMP data, size ", input.size()));
+  }
+
+  uint8* Decode(const void* srcdata, int datasize,
+              std::function<uint8*(int, int, int)> allocate_output);
+};
+REGISTER_KERNEL_BUILDER(Name("DecodeBmp").Device(DEVICE_CPU), DecodeBmpOp);
+
+uint8* DecodeBmpOp::Decode(const void* srcdata, int datasize,
+              std::function<uint8*(int, int, int)> allocate_output) {
+
+  bool top_down = false;
+  uint8* const img_bytes = (uint8 *) srcdata;
+  const int header_size = *(reinterpret_cast<int*>(img_bytes + 10));
+  const int width = *(reinterpret_cast<int*>(img_bytes + 18));
+  const int height = *(reinterpret_cast<int*>(img_bytes + 22));
+
+  // if height is negative, data layout is top down
+  // otherwise, it's bottom up
+  if (height < 0)
+    top_down = true;
+  uint8* dstdata = allocate_output(width, abs(height), 3);
+  uint8* const bmp_pixels = &img_bytes[header_size];
+  int row_size = (8 * width * 3 + 31) / 32 * 4;
+
+#ifndef NDEBUG
+  const int data_size = *(reinterpret_cast<int*>(img_bytes + 34));
+  LOG(INFO) << "header size = " << header_size;
+  LOG(INFO) << "width = " << width;
+  LOG(INFO) << "height = " << height;
+  LOG(INFO) << "data size = " << data_size;
+  LOG(INFO) << "row size = " << row_size;
+  LOG(INFO) << "top down = " << top_down;
+#endif
+
+  for (int i = 0; i < abs(height); i++) {
```

Comment:
Awesome, this is probably sufficient to me to move forward -- it's possible it could be made faster, but as long as it isn't a computational bottleneck for real models, it seems okay.  

Do you have the real numbers to share of running this?